pkg/filters/proxy/wsproxy.go

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package proxy
+
+import (
+	"fmt"
+
+	"github.com/megaease/easegress/pkg/context"
+	"github.com/megaease/easegress/pkg/filters"
+	"github.com/megaease/easegress/pkg/protocols/httpprot"
+	"github.com/megaease/easegress/pkg/supervisor"
+)
+
+// WebSocketProxyKind is the kind of WebSocketProxy.
+const WebSocketProxyKind = "WebSocketProxy"
+
+var kindWebSocketProxy = &filters.Kind{
```

Comment:
IMO, it's better to split every filter into different directories even they have the same infrastructure, which makes the layout clean and has no particular case.

Reply:
because the two filters share a lot of code, and the limit time for implement this feature, we have to balance between:
* put them in the same folder
* duplicate the code

and I think the 2nd is worse.

we can do a refactor to put them into different folders later, this only require code level changes, and won't impact the external interface, so we can do it at any time.

As I mentioned before, because we are now using Go 1.18, it is possible for us to extract the load balance and request matching stuff into seperate packages, and I think it will be easier for us to do the refactor after that.

Reply:
Your argument makes sense to me. The only thing I was curious about is why different folders mean duplicating code with exporting infrastructure.

Reply:
the shared code including:
* load balance
* request matching
* service discovery

if I put the filters into different folders, and avoid code duplication at the same time, I need to extract them into seperate packages, but, I don't have enough time to make a good design for the interfaces of the new packages (in fact, I already did some refactor towards this, e.g. extract service discovery to the `BaseServerPool` struct).

Reply:
If we compromise this for time cost, I think we should add some TODO comments in the implementation to remind us to refactor it in the future.

Reply:
sure, added.

Reply:
I see, the real problem is no time to design interfaces instead of inevitable duplicate code.